src/float.rs

```diff
@@ -1806,6 +1806,37 @@ pub trait Float: Num + Copy + NumCast + PartialOrd + Neg<Output = Self> {
     /// assert!(abs_difference < 1e-10);
     /// ```
     fn integer_decode(self) -> (u64, i16, i8);
+
+    /// Returns a number composed of the magnitude of `self` and the sign of
+    /// `sign`.
+    ///
+    /// Equal to `self` if the sign of `self` and `sign` are the same, otherwise
+    /// equal to `-self`. If `self` is a `NAN`, then a `NAN` with the sign of
+    /// `sign` is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use num_traits::Float;
+    ///
+    /// let f = 3.5_f32;
+    ///
+    /// assert_eq!(f.copysign(0.42), 3.5_f32);
+    /// assert_eq!(f.copysign(-0.42), -3.5_f32);
+    /// assert_eq!((-f).copysign(0.42), 3.5_f32);
+    /// assert_eq!((-f).copysign(-0.42), -3.5_f32);
+    ///
+    /// assert!(f32::nan().copysign(1.0).is_nan());
+    /// ```
+    fn copysign(self, sign:Self) -> Self {
+        if self.is_nan() {
+            sign / Self::zero()
+        } else if self.signum() == sign.signum() {
```

Comment:
`is_sign_negative` is better than `signum` here as it works on NaNs as well; `signum` itself is implemented using `is_sign_negative` in the trait.

Then the special case for NaN can be removed as both `is_sign_negative` and `neg` work well on NaNs. (Currently the NaN special case is buggy anyway and returns infinity rather than NaN.)